File: account/src/main/java/serp/project/account/ui/controller/DepartmentController.java
/**
 * Author: QuanTuanHuy
 * Description: Part of Serp Project
 */

package serp.project.account.ui.controller;

import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import serp.project.account.core.domain.constant.Constants;
import serp.project.account.core.domain.dto.request.AssignUserToDepartmentRequest;
import serp.project.account.core.domain.dto.request.CreateDepartmentRequest;
import serp.project.account.core.domain.dto.request.GetDepartmentParams;
import serp.project.account.core.domain.dto.request.UpdateDepartmentRequest;
import serp.project.account.core.usecase.DepartmentUseCase;
import serp.project.account.kernel.utils.AuthUtils;
import serp.project.account.kernel.utils.ResponseUtils;

@RequiredArgsConstructor
@RestController
@RequestMapping("/api/v1/organizations/{organizationId}/departments")
@Slf4j
public class DepartmentController {
    private final DepartmentUseCase departmentUseCase;
    private final AuthUtils authUtils;
    private final ResponseUtils responseUtil;

    @PostMapping
    public ResponseEntity<?> createDepartment(
            @PathVariable Long organizationId,
            @Valid @RequestBody CreateDepartmentRequest request) {
        if (!authUtils.canAccessOrganization(organizationId)) {
            var resp = responseUtil.forbidden(Constants.ErrorMessage.NO_PERMISSION_TO_ACCESS_ORGANIZATION);
            return ResponseEntity.status(resp.getCode()).body(resp);
        }

        var response = departmentUseCase.createDepartment(organizationId, request);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @PatchMapping("/{departmentId}")
    public ResponseEntity<?> updateDepartment(
            @PathVariable Long organizationId,
            @PathVariable Long departmentId,
            @Valid @RequestBody UpdateDepartmentRequest request) {
        if (!authUtils.canAccessOrganization(organizationId)) {
            var resp = responseUtil.forbidden(Constants.ErrorMessage.NO_PERMISSION_TO_ACCESS_ORGANIZATION);
            return ResponseEntity.status(resp.getCode()).body(resp);
        }
        var response = departmentUseCase.updateDepartment(departmentId, request);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @DeleteMapping("/{departmentId}")
    public ResponseEntity<?> deleteDepartment(
            @PathVariable Long organizationId,
            @PathVariable Long departmentId) {
        if (!authUtils.canAccessOrganization(organizationId)) {
            var resp = responseUtil.forbidden(Constants.ErrorMessage.NO_PERMISSION_TO_ACCESS_ORGANIZATION);
            return ResponseEntity.status(resp.getCode()).body(resp);
        }
        var response = departmentUseCase.deleteDepartment(departmentId);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @GetMapping("/{departmentId}")
    public ResponseEntity<?> getDepartmentById(
            @PathVariable Long organizationId,
            @PathVariable Long departmentId) {

        if (!authUtils.canAccessOrganization(organizationId)) {
            var resp = responseUtil.forbidden(Constants.ErrorMessage.NO_PERMISSION_TO_ACCESS_ORGANIZATION);
            return ResponseEntity.status(resp.getCode()).body(resp);
        }

        var response = departmentUseCase.getDepartmentById(departmentId);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @GetMapping
    public ResponseEntity<?> getDepartments(
            @PathVariable Long organizationId,
            @RequestParam(required = false) Integer page,
            @RequestParam(required = false) Integer pageSize,
            @RequestParam(required = false) String sortBy,
            @RequestParam(required = false) String sortDir,
            @RequestParam(required = false) String search,
            @RequestParam(required = false) Long parentDepartmentId,
            @RequestParam(required = false) Boolean isActive,
            @RequestParam(required = false) Long managerId) {

        if (!authUtils.canAccessOrganization(organizationId)) {
            var resp = responseUtil.forbidden(Constants.ErrorMessage.NO_PERMISSION_TO_ACCESS_ORGANIZATION);
            return ResponseEntity.status(resp.getCode()).body(resp);
        }

        GetDepartmentParams params = GetDepartmentParams.builder()
                .page(page)
                .pageSize(pageSize)
                .sortBy(sortBy)
                .sortDirection(sortDir)
                .search(search)
                .organizationId(organizationId)
                .parentDepartmentId(parentDepartmentId)
                .isActive(isActive)
                .managerId(managerId)
                .build();

        var response = departmentUseCase.getDepartmentsByOrganization(params);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @PostMapping("/{departmentId}/users")
    public ResponseEntity<?> assignUserToDepartment(
            @PathVariable Long organizationId,
            @PathVariable Long departmentId,
            @Valid @RequestBody AssignUserToDepartmentRequest request) {

        if (!authUtils.canAccessOrganization(organizationId)) {
            var resp = responseUtil.forbidden(Constants.ErrorMessage.NO_PERMISSION_TO_ACCESS_ORGANIZATION);
            return ResponseEntity.status(resp.getCode()).body(resp);
        }

        request.setDepartmentId(departmentId);

        var response = departmentUseCase.assignUserToDepartment(request);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @GetMapping("/{departmentId}/members")
    public ResponseEntity<?> getDepartmentMembers(
            @PathVariable Long organizationId,
            @PathVariable Long departmentId) {

        if (!authUtils.canAccessOrganization(organizationId)) {
            var resp = responseUtil.forbidden(Constants.ErrorMessage.NO_PERMISSION_TO_ACCESS_ORGANIZATION);
            return ResponseEntity.status(resp.getCode()).body(resp);
        }

        var response = departmentUseCase.getMembersByDepartmentId(departmentId);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @GetMapping("/stats")
    public ResponseEntity<?> getDepartmentStats(
            @PathVariable Long organizationId) {
        if (!authUtils.canAccessOrganization(organizationId)) {
            var resp = responseUtil.forbidden(Constants.ErrorMessage.NO_PERMISSION_TO_ACCESS_ORGANIZATION);
            return ResponseEntity.status(resp.getCode()).body(resp);
        }
        var response = departmentUseCase.getDepartmentStats(organizationId);
        return ResponseEntity.status(response.getCode()).body(response);
    }

}

File: api_gateway/src/ui/controller/account/department_controller.go
/*
Author: QuanTuanHuy
Description: Part of Serp Project
*/

package controller

import (
	"github.com/gin-gonic/gin"
	"github.com/serp/api-gateway/src/core/domain/constant"
	service "github.com/serp/api-gateway/src/core/service/account"
	"github.com/serp/api-gateway/src/kernel/utils"
)

type DepartmentController struct {
	deptService service.IDepartmentService
}

func (d *DepartmentController) CreateDepartment(c *gin.Context) {
	orgId, ok := utils.ValidateAndParseID(c, "organizationId")
	if !ok {
		return
	}
	var body map[string]any
	if err := c.ShouldBindJSON(&body); err != nil {
		utils.AbortErrorHandle(c, constant.GeneralBadRequest)
		return
	}
	res, err := d.deptService.CreateDepartment(c.Request.Context(), orgId, body)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (d *DepartmentController) GetDepartments(c *gin.Context) {
	orgId, ok := utils.ValidateAndParseID(c, "organizationId")
	if !ok {
		return
	}

	q := map[string]string{}
	if v := c.Query("page"); v != "" {
		q["page"] = v
	}
	if v := c.Query("pageSize"); v != "" {
		q["pageSize"] = v
	}
	if v := c.Query("sortBy"); v != "" {
		q["sortBy"] = v
	}
	if v := c.Query("sortDir"); v != "" {
		q["sortDir"] = v
	}
	if v := c.Query("search"); v != "" {
		q["search"] = v
	}
	if v := c.Query("parentDepartmentId"); v != "" {
		q["parentDepartmentId"] = v
	}
	if v := c.Query("isActive"); v != "" {
		q["isActive"] = v
	}
	if v := c.Query("managerId"); v != "" {
		q["managerId"] = v
	}

	res, err := d.deptService.GetDepartments(c.Request.Context(), orgId, q)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (d *DepartmentController) GetDepartmentById(c *gin.Context) {
	orgId, ok := utils.ValidateAndParseID(c, "organizationId")
	if !ok {
		return
	}
	departmentId, ok := utils.ValidateAndParseID(c, "departmentId")
	if !ok {
		return
	}
	res, err := d.deptService.GetDepartmentById(c.Request.Context(), orgId, departmentId)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (d *DepartmentController) UpdateDepartment(c *gin.Context) {
	orgId, ok := utils.ValidateAndParseID(c, "organizationId")
	if !ok {
		return
	}
	departmentId, ok := utils.ValidateAndParseID(c, "departmentId")
	if !ok {
		return
	}
	var body map[string]any
	if err := c.ShouldBindJSON(&body); err != nil {
		utils.AbortErrorHandle(c, constant.GeneralBadRequest)
		return
	}
	res, err := d.deptService.UpdateDepartment(c.Request.Context(), orgId, departmentId, body)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (d *DepartmentController) DeleteDepartment(c *gin.Context) {
	orgId, ok := utils.ValidateAndParseID(c, "organizationId")
	if !ok {
		return
	}
	departmentId, ok := utils.ValidateAndParseID(c, "departmentId")
	if !ok {
		return
	}
	res, err := d.deptService.DeleteDepartment(c.Request.Context(), orgId, departmentId)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (d *DepartmentController) GetDepartmentTree(c *gin.Context) {
	orgId, ok := utils.ValidateAndParseID(c, "organizationId")
	if !ok {
		return
	}
	departmentId, ok := utils.ValidateAndParseID(c, "departmentId")
	if !ok {
		return
	}
	res, err := d.deptService.GetDepartmentTree(c.Request.Context(), orgId, departmentId)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (d *DepartmentController) AssignUserToDepartment(c *gin.Context) {
	orgId, ok := utils.ValidateAndParseID(c, "organizationId")
	if !ok {
		return
	}
	departmentId, ok := utils.ValidateAndParseID(c, "departmentId")
	if !ok {
		return
	}
	var body map[string]interface{}
	if err := c.ShouldBindJSON(&body); err != nil {
		utils.AbortErrorHandle(c, constant.GeneralBadRequest)
		return
	}
	res, err := d.deptService.AssignUserToDepartment(c.Request.Context(), orgId, departmentId, body)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (d *DepartmentController) BulkAssignUsersToDepartment(c *gin.Context) {
	orgId, ok := utils.ValidateAndParseID(c, "organizationId")
	if !ok {
		return
	}
	departmentId, ok := utils.ValidateAndParseID(c, "departmentId")
	if !ok {
		return
	}
	var body map[string]any
	if err := c.ShouldBindJSON(&body); err != nil {
		utils.AbortErrorHandle(c, constant.GeneralBadRequest)
		return
	}
	res, err := d.deptService.BulkAssignUsersToDepartment(c.Request.Context(), orgId, departmentId, body)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (d *DepartmentController) RemoveUserFromDepartment(c *gin.Context) {
	orgId, ok := utils.ValidateAndParseID(c, "organizationId")
	if !ok {
		return
	}
	departmentId, ok := utils.ValidateAndParseID(c, "departmentId")
	if !ok {
		return
	}
	userId, ok := utils.ValidateAndParseID(c, "userId")
	if !ok {
		return
	}
	res, err := d.deptService.RemoveUserFromDepartment(c.Request.Context(), orgId, departmentId, userId)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (d *DepartmentController) GetDepartmentMembers(c *gin.Context) {
	orgId, ok := utils.ValidateAndParseID(c, "organizationId")
	if !ok {
		return
	}
	departmentId, ok := utils.ValidateAndParseID(c, "departmentId")
	if !ok {
		return
	}
	res, err := d.deptService.GetDepartmentMembers(c.Request.Context(), orgId, departmentId)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (d *DepartmentController) GetDepartmentStats(c *gin.Context) {
	orgId, ok := utils.ValidateAndParseID(c, "organizationId")
	if !ok {
		return
	}
	res, err := d.deptService.GetDepartmentStats(c.Request.Context(), orgId)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func NewDepartmentController(deptService service.IDepartmentService) *DepartmentController {
	return &DepartmentController{deptService: deptService}
}
